modules/peridynamics/src/kernels/HorizonStabilizedFormIIFiniteStrainMechanicsNOSPD.C

```diff
@@ -0,0 +1,503 @@
+{
+  // includes dTi/dUj and dTj/dUi contributions
+  for (unsigned int nd = 0; nd < _nnodes; ++nd)
+  {
+    RankFourTensor dSdFhat = computeDSDFhat(nd);
+    RankTwoTensor invF = _dgrad[nd].inverse();
+    Real detF = _dgrad[nd].det();
+    // calculation of jacobian contribution to current_node's neighbors
+    std::vector<dof_id_type> ivardofs(_nnodes), jvardofs(_nnodes);
+    ivardofs[0] = _current_elem->node_ptr(nd)->dof_number(_sys.number(), _var.number(), 0);
+    jvardofs[0] = _current_elem->node_ptr(nd)->dof_number(_sys.number(), _var.number(), 0);
+    std::vector<dof_id_type> neighbors = _pdmesh.getNeighbors(_current_elem->node_id(nd));
+    std::vector<dof_id_type> bonds = _pdmesh.getBonds(_current_elem->node_id(nd));
+
+    dof_id_type nb_index =
+        std::find(neighbors.begin(), neighbors.end(), _current_elem->node_id(1 - nd)) -
+        neighbors.begin();
+    std::vector<dof_id_type> dg_neighbors =
+        _pdmesh.getBondDeformationGradientNeighbors(_current_elem->node_id(nd), nb_index);
+
+    RealGradient origin_vec_nb1;
+
+    for (unsigned int nb1 = 0; nb1 < dg_neighbors.size(); ++nb1)
+      if (_bond_status_var->getElementalValue(_pdmesh.elemPtr(bonds[dg_neighbors[nb1]])) > 0.5)
+      {
+        ivardofs[1] = _pdmesh.nodePtr(neighbors[dg_neighbors[nb1]])
+                          ->dof_number(_sys.number(), _var.number(), 0);
+        origin_vec_nb1 = *_pdmesh.nodePtr(neighbors[dg_neighbors[nb1]]) -
+                         *_pdmesh.nodePtr(_current_elem->node_id(nd));
+
+        Real vol_nb2, dJdU;
+        RealGradient origin_vec_nb2;
+        RankTwoTensor dFdUk, dPxdUkx, dSdU, dinvFTdU;
+
+        for (unsigned int nb2 = 0; nb2 < dg_neighbors.size(); ++nb2)
+          if (_bond_status_var->getElementalValue(_pdmesh.elemPtr(bonds[dg_neighbors[nb2]])) > 0.5)
+          {
+            ivardofs[1] = _pdmesh.nodePtr(neighbors[dg_neighbors[nb2]])
+                              ->dof_number(_sys.number(), _var.number(), 0);
+            vol_nb2 = _pdmesh.getNodeVolume(neighbors[dg_neighbors[nb2]]);
+
+            origin_vec_nb2 = *_pdmesh.nodePtr(neighbors[dg_neighbors[nb2]]) -
+                             *_pdmesh.nodePtr(_current_elem->node_id(nd));
+
+            dFdUk.zero();
+            for (unsigned int i = 0; i < _dim; ++i)
+              dFdUk(_component, i) =
+                  _horizon_radius[nd] / origin_vec_nb2.norm() * origin_vec_nb2(i) * vol_nb2;
+
+            dFdUk *= _shape2[nd].inverse();
+
+            // calculate dJ/du
+            dJdU = 0.0;
+            for (unsigned int i = 0; i < 3; ++i)
+              for (unsigned int j = 0; j < 3; ++j)
+                dJdU += detF * invF(j, i) * dFdUk(i, j);
+
+            // calculate dS/du
+            dSdU = dSdFhat * dFdUk * _dgrad_old[nd].inverse();
+
+            // calculate dinv(F)Tdu
+            dinvFTdU.zero();
+            for (unsigned int i = 0; i < 3; ++i)
+              for (unsigned int J = 0; J < 3; ++J)
```

Comment:
Using upper and lower case variable names for indices is a little odd. It seems like J and L can be lower case here without conflicting with anything.